### ISSUE TITLE
Structuring Transfer Learning

### ISSUE THREAD

**Issue by vsomnath** [deepchem/deepchem#1906]
Starting off from Peter's Transfer Learning comments Peter's in #1834, 
> For the transfer learning code, I wonder if we want to make it a bit more generic? As currently written, it feels more like an example than a feature. For example, it's hardcoded to pretrain on one specific dataset. 

The PR included only the ChemNet specific protocol for transfer learning, but it would be useful to have a more general framework. To the best of my knowledge, transfer learning for molecules/chemistry falls along three axes:

1. ChemNet Transfer Learning protocol (https://arxiv.org/abs/1712.02734): Supervised with auxiliary labels
2. BERT Style pretraining using SMILES strings (https://huggingface.co/seyonec/ChemBERTa-zinc-base-v1) : Self-supervised
3. Graph-Based Pretraining (https://openreview.net/forum?id=HJlWWJSFDH) - Self-supervised 

Using this issue to kickoff discussions on this.

**Comment by rbharath** [comment]
Thanks for kicking off this discussion! I think a couple other variants to consider are 

4. Doing contrastive learning on ChemNet style images. That is, using standard image constrastive learning with chemnet style encoding of images.
5. Using multiple conformations as a data augmentation technique for self-supervised learning as suggested in this forum post (https://forum.deepchem.io/t/molecular-attention-transformers/91/5?u=bharath)

I'm not sure yet what a generic pretraining/transfer learning API should look like. One challenge is that different pretraining methods are tied to different models. ChemNet Transfer is tied to ChemNet (although you can imagine RDKit property pretraining as a general variant), BERT pretraining is tied to transformers, graph-based pretraining to graph convolutions, etc.

Perhaps we can add `pretrain()` as a non-implemented method to the abstract superclass `Model`. Then different subclasses can override this with their own custom pretraining method. This allows different models to support different pretraining modes.

**Comment by peastman** [comment]
What exactly would `pretrain()` do?  It would be useful to define the scope of this feature, and what problems it's intended to solve.  What do we really mean by pretraining, beyond simply calling `fit()` twice with two different datasets?

**Comment by rbharath** [comment]
I think the main difference would be that `pretrain()` could have a different objective function than `fit()`. I think for ChemBERTa style models for example, the `pretrain()` objective would be masked language modeling, while the `fit()` object function would be a standard L2loss or Cross-entropy loss.

Perhaps one way of implementing this would be to allow a model to have multiple objective functions and allow `fit()` to select between them. Then `pretrain()` could just be a convenience wrapper around a suitable `fit()` call.

**Comment by peastman** [comment]
> Perhaps one way of implementing this would be to allow a model to have multiple objective functions and allow fit() to select between them.

`fit()` already allows you to provide an alternate loss function, so that would be very easy.

Is that the only difference?

**Comment by rbharath** [comment]
Oh wow, didn't notice we already had that feature!

Hmm, that might actually be the only major difference. I think other differences between pretraining and regular training are potentially different datasets, different transformers, different metrics, but these are all outside the core training loop. 

Perhaps we could even have a generic implementation for `pretrain()` as follows. If `self._pretrain_loss_fn` is set for a model, then we implement

```
def pretrain(self, dataset, **kwargs):
  if not self._pretrain_loss_fn:
    raise ValueError("Must set pretraining loss to pretrain a model.")
  else:
    self.fit(dataset, loss=self._pretrain_loss_fn, **kwargs)
```

@vsomnath @seyonechithrananda Would this design fail to work for transfer learning methods you've used?

**Comment by rbharath** [comment]
A second change I'd propose is that we add a `load_pretrained()` method to `Model`. This method would be akin to the `dc.molnet.load_X` methods in that it would query S3 for a set of pretrained weights for this model class and see if there's a set of weights that corresponds to the hyperparameters specified for the class. This would be useful since many pretraining methods are exorbitantly expensive and having the ability to load pretrained weights would make a big difference for usersl

**Comment by vsomnath** [comment]
@rbharath: This would work just fine for ChemNet. 
For the `load_pretrained`, I had setup a similar [configuration](https://github.com/deepchem/deepchem/blob/618d0b2bfd643eee7162635cb10da58fad46f55c/deepchem/transfer_learning/chemnet.py) but having it as a class method would be cleaner.

**Comment by seyonechithrananda** [comment]
> A second change I'd propose is that we add a `load_pretrained()` method to `Model`. This method would be akin to the `dc.molnet.load_X` methods in that it would query S3 for a set of pretrained weights for this model class and see if there's a set of weights that corresponds to the hyperparameters specified for the class. This would be useful since many pretraining methods are exorbitantly expensive and having the ability to load pre-trained weights would make a big difference for users

For ChembERTa, this could work well for creating a model instance using HuggingFace's model hub which is stored in S3. Then, users can fine-tune on sequence classification from there on a dataset of their choice.

**Comment by peastman** [comment]
> This method would be akin to the dc.molnet.load_X methods in that it would query S3 for a set of pretrained weights for this model class and see if there's a set of weights that corresponds to the hyperparameters specified for the class.

I would do it a little differently.  That puts the burden on the user to know the exact values of all hyperparameters for the pretrained model they want, then create a model with exactly those values.  Instead let's have a function that constructs and returns a new model object.

It could even be a single function for all models.  You provide an identifier that specifies the model you want.  It looks up that identifier online and retrieves the information needed to create it.  That would include the name of the model class to create; the values of all arguments to pass to the constructor; and a checkpoint to load the parameters from.

**Comment by seyonechithrananda** [comment]
> > This method would be akin to the dc.molnet.load_X methods in that it would query S3 for a set of pretrained weights for this model class and see if there's a set of weights that corresponds to the hyperparameters specified for the class.
> 
> I would do it a little differently. That puts the burden on the user to know the exact values of all hyperparameters for the pretrained model they want, then create a model with exactly those values. Instead let's have a function that constructs and returns a new model object.
> 
> It could even be a single function for all models. You provide an identifier that specifies the model you want. It looks up that identifier online and retrieves the information needed to create it. That would include the name of the model class to create; the values of all arguments to pass to the constructor; and a checkpoint to load the parameters from.

I think this is a really good idea. It functions well in practice and has the added ease of use, similar to how Huggingface does it. They allow users to get the config (which contains hyperparameters), tokenizer and model all from the same identifier:

```
tokenizer = AutoTokenizer.from_pretrained('distilgpt2')   # Initialize tokenizer
model = AutoModelWithLMHead.from_pretrained('distilgpt2')    # Download model and configuration from S3 and cache.
outputs = model.generate(max_length=40)  # do greedy decoding
print('Generated: {}'.format(tokenizer.decode(outputs[0], skip_special_tokens=True)))

```

**Comment by vsomnath** [comment]
> It could even be a single function for all models. You provide an identifier that specifies the model you want. It looks up that identifier online and retrieves the information needed to create it. That would include the name of the model class to create; the values of all arguments to pass to the constructor; and a checkpoint to load the parameters from.

The way I had the current protocol up was to construct a model with any user-specified hyperparameters. An identifier based on these hyperparameters is constructed and queried against a list of identifiers for which pretrained models are available. If a match exists, then you download and load that pretrained model, else skip it.

**Comment by peastman** [comment]
Did you intend to close this?

**Comment by rbharath** [comment]
Oh oops! Sorry, I was switching tabs and accidentally clicked close. My bad

**Comment by rbharath** [comment]
This has been implemented by the new `ModularTorchModel` framework. Going to close this old discussion
